add logging with using proxy pattern over interface implementation

freegate78 opened the pull request:
> Task :HW09-ByteCodes:Main.main()
invoking method:public abstract void com.akproductions.testlogging.TestLoggingInterface.calculation(int)

freegate78 committed:
add logging with using proxy pattern over interface implementation

> Task :HW09-ByteCodes:Main.main()
invoking method:public abstract void com.akproductions.testlogging.TestLoggingInterface.calculation(int)

freegate78 committed:
добавил в аннотационный интерфейс retention runtime
добавил аннотацию @Log в интерфейс, который оборачиваем в прокси
добавил в проксированный метод invoke анализ аннотации @Log (заметил что аннотация видно только на интерфейсе, на исплементации не видна, т.к. работает с классом интерфейса, а у него только метод есть в доступных. без интерфейса нужен чистый asm на методе, лень :) )

итого условное  логирование заработало
[LOG] invoking method:public abstract void com.akproductions.testlogging.TestLoggingInterface.calculation(int) with args: 10
called method calculation with param 10

kataus commented on HW09-ByteCodes/src/main/java/com/akproductions/proxycalldemo/Ioc.java:
```diff
@@ -0,0 +1,51 @@
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            Annotation[] annotations = method.getDeclaredAnnotations();
```
Рефлексия достаточно дорогое удовольствие в работе. Дергать её на каждый вызов - излишние накладные расходы. давайте сделаем чтобы только один раз за время жизни мы дергали определение логировать ли класс. Оно неизменно и нет особого смысла каждый раз проверять

freegate78 committed:
1) преобразовал класс результатов по рекомендации - не HashMap, а конкретные метрики тестов, а также методы для их установки и изменения
2) исправил параметры методов на интерфейс map <Method>

kataus commented on HW09-ByteCodes/src/main/java/com/akproductions/proxycalldemo/Ioc.java:
```diff
@@ -29,15 +28,11 @@ public static TestLoggingInterface createMyClass() {
-            for (Annotation annotation : annotations) {
-                if (annotation.annotationType().equals(Log.class))
+
+            if(method.isAnnotationPresent(Log.class))
```
Последнее осталось. Каждый раз запрашивать рефлексию достаточно дорого, давайте сделаем один раз при запуске определение методов для логгирования и будем уже с коротким списком сверяться, а не по рефлексии определять

freegate78 committed:
добавил в класс ioc константу - для определения надо ли логировать
  static final boolean EnableLogging = com.akproductions.testlogging.TestLoggingInterface.class.getDeclaredMethods()[0].isAnnotationPresent(Log.class);

kataus commented on HW09-ByteCodes/src/main/java/com/akproductions/proxycalldemo/Ioc.java:
```diff
@@ -0,0 +1,41 @@
+import java.lang.reflect.Proxy;
+
+public class Ioc {
+    static final boolean EnableLogging = com.akproductions.testlogging.TestLoggingInterface.class.getDeclaredMethods()[0].isAnnotationPresent(Log.class);
```
Это работает только если первый метод помечен аннотацией @Log. А остальные? Давайте сделаем в запуске экземпляра класса полный скан и сделаем Set или Map со списоком методов которые должны логгироваться. И при invoke просто быстро проверять наличие метода в списке.

freegate78 committed:
сделал метод
    private static List<Class> getLoggedMethods(Method[] methods, Class classOfAnnotation)
вызываю его в createmyclass
он заполняет аррайлист
который уже дергается при инвоке

freegate78 committed:
строчка лишняя была убрал дебажил с ней
        varLoggedMethods.add(classOfAnnotation);

kataus commented on HW09-ByteCodes/src/main/java/com/akproductions/proxycalldemo/Ioc.java:
```diff
@@ -0,0 +1,55 @@
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            if(LoggedMethods.contains(method.getClass())) {System.out.println("[LOG] invoking method:" + method + " with args: "+ args[0]);}
```
Не стоит делать такие длинные строки. тут можно легко разделить условие и действие
